### PR TITLE
GLT-2244: allow bulk creating orders with different parameters

### PR DESCRIPTION
**Pull request by djcooke**

### CODE REVIEW

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditOrderController.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      return dto;
+    }
+
+    @Override
+    protected Stream<Pool> loadParents(List<Long> parentIds) throws IOException {
+      return poolService.listByIdList(parentIds).stream();
+    }
+
+    @Override
+    protected void writeConfiguration(ObjectMapper mapper, ObjectNode config) throws IOException {
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/EditOrderController.java]: CRITICAL Add a nested comment explaining why this method is empty, throw an UnsupportedOperationException or complete the implementation. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1186)

File: miso-web/src/main/webapp/scripts/hot_order.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            var platform = Utils.array.findFirstOrNull(function(platform) {
+              return platform.instrumentModel == value;
+            }, Constants.platforms);
+            var params = Constants.sequencingParameters.filter(function(parameters) {
+              return parameters.platform.id == platform.id;
+            }).map(function(parameters) {
+              return parameters.name;
+            }).sort();
+          } else {
+            var params = [];
```
Comment (sqwci) [inline miso-web/src/main/webapp/scripts/hot_order.js]: MAJOR Rename "params" as this name is already used in declaration at line 80. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2814)

**Comment by sqwci** [comment]
SonarQube analysis reported 8 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [PoolOrderRestController.java#L142](https://github.com/TGAC/miso-lims/blob/82336067296d246bb8f9791489c0350e16ad7c22/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/PoolOrderRestController.java#L142): Remove usage of generic wildcard type. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1452)
1. ![MAJOR][MAJOR] [Dtos.java#L153](https://github.com/TGAC/miso-lims/blob/82336067296d246bb8f9791489c0350e16ad7c22/miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java#L153): Add a private constructor to hide the implicit public one. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1118)
1. ![MAJOR][MAJOR] [Dtos.java#L711](https://github.com/TGAC/miso-lims/blob/82336067296d246bb8f9791489c0350e16ad7c22/miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java#L711): Remove this unused method parameter "from". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [Dtos.java#L1019](https://github.com/TGAC/miso-lims/blob/82336067296d246bb8f9791489c0350e16ad7c22/miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java#L1019): Make this anonymous inner class a lambda [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1604)
1. ![MINOR][MINOR] [Dtos.java#L712](https://github.com/TGAC/miso-lims/blob/82336067296d246bb8f9791489c0350e16ad7c22/miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java#L712): Immediately return this expression instead of assigning it to the temporary variable "to". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1488)
1. ![MINOR][MINOR] [DefaultPoolOrderService.java#L80](https://github.com/TGAC/miso-lims/blob/82336067296d246bb8f9791489c0350e16ad7c22/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultPoolOrderService.java#L80): Remove the declaration of thrown exception 'uk.ac.bbsrc.tgac.miso.service.security.AuthorizationException' which is a runtime exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: miso-web/src/main/webapp/scripts/hot_order.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+          }
+          setOptions({
+            source: params
+          });
+        }
+      }, HotUtils.makeColumnForInt('Partitions', true, 'partitions', HotUtils.validator.requiredNumber)];
```
Comment (callunity) [inline miso-web/src/main/webapp/scripts/hot_order.js]: Hmm, we've been able to enter negative numbers on all our `requiredNumber`s this whole time... so looks like it's not an issue in production, but might be nice to guard against negative numbers in the validator.

Comment (djcooke) [reply miso-web/src/main/webapp/scripts/hot_order.js]: `requiredNumber` allowed all kinds of numbers, but in all cases where we used it, we actually wanted positive integers only. I changed it to `requiredPositiveInt`

File: miso-web/src/main/webapp/scripts/hot_sample.js [hunk 4]
```diff
@@ -584,93 +585,91 @@ HotTarget.sample = (function() {
...
-                    }
-                  };
+      return [{
+        name: "Edit",
+        action: function(samples) {
+
+          if (samples.some(function(sample) {
+            return sample.sampleClassId;
+          }) && !Constants.isDetailedSample) {
+            alert("There's detailed samples, but MISO is not configured for this.");
```
Comment (sqwci) [inline miso-web/src/main/webapp/scripts/hot_sample.js]: MINOR Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)

File: miso-web/src/main/webapp/scripts/hot_sample.js [hunk 5]
```diff
@@ -584,93 +585,91 @@ HotTarget.sample = (function() {
...
-                  return !!x;
-                }), function(result) {
-                  (result.target || targets[0]).action(result.replicates);
-                });
               });
             }
-          }, HotUtils.printAction('sample'), HotUtils.spreadsheetAction('/miso/rest/sample/spreadsheet', Constants.sampleSpreadsheets),
+
+            if (targets.length == 0) {
+              alert("No propagation is possible from the samples.");
```
Comment (sqwci) [inline miso-web/src/main/webapp/scripts/hot_sample.js]: MINOR Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)

File: miso-web/src/main/webapp/scripts/hot_sample.js [hunk 6]
```diff
@@ -584,93 +585,91 @@ HotTarget.sample = (function() {
...
+            alert("There's detailed samples, but MISO is not configured for this.");
+            return;
+          }
+
+          var classes = getSampleClasses(samples);
+          var classesAliases = Utils.array.deduplicateString(classes.map(function(sampleClass) {
+            return sampleClass.alias;
+          }));
+          if (classesAliases.length > 1) {
+            alert("You have selected samples of classes " + classesAliases.join(" & ") + ". Please select samples from only one class.");
```
Comment (sqwci) [inline miso-web/src/main/webapp/scripts/hot_sample.js]: MINOR Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)

**Comment by sqwci** [comment]
SonarQube analysis reported 19 issues
* ![BLOCKER][BLOCKER] 3 blocker
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 9 major
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [hot.js#L336](https://github.com/TGAC/miso-lims/blob/73862dfebeda7ae6c1c1fc7017ed296a57eaa7b8/miso-web/src/main/webapp/scripts/hot.js#L336): Add the "let", "const" or "var" keyword to this declaration of "prop" to make it explicit. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2703)
1. ![BLOCKER][BLOCKER] [hot.js#L381](https://github.com/TGAC/miso-lims/blob/73862dfebeda7ae6c1c1fc7017ed296a57eaa7b8/miso-web/src/main/webapp/scripts/hot.js#L381): Add the "let", "const" or "var" keyword to this declaration of "needsRender" to make it explicit. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2703)
1. ![BLOCKER][BLOCKER] [hot.js#L1000](https://github.com/TGAC/miso-lims/blob/73862dfebeda7ae6c1c1fc7017ed296a57eaa7b8/miso-web/src/main/webapp/scripts/hot.js#L1000): Add the "let", "const" or "var" keyword to this declaration of "HotTarget" to make it explicit. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2703)
1. ![CRITICAL][CRITICAL] [PoolOrderRestController.java#L142](https://github.com/TGAC/miso-lims/blob/73862dfebeda7ae6c1c1fc7017ed296a57eaa7b8/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/rest/PoolOrderRestController.java#L142): Remove usage of generic wildcard type. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1452)
1. ![MAJOR][MAJOR] [Dtos.java#L153](https://github.com/TGAC/miso-lims/blob/73862dfebeda7ae6c1c1fc7017ed296a57eaa7b8/miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java#L153): Add a private constructor to hide the implicit public one. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1118)
1. ![MAJOR][MAJOR] [Dtos.java#L711](https://github.com/TGAC/miso-lims/blob/73862dfebeda7ae6c1c1fc7017ed296a57eaa7b8/miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java#L711): Remove this unused method parameter "from". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1172)
1. ![MAJOR][MAJOR] [Dtos.java#L1019](https://github.com/TGAC/miso-lims/blob/73862dfebeda7ae6c1c1fc7017ed296a57eaa7b8/miso-dto/src/main/java/uk/ac/bbsrc/tgac/miso/dto/Dtos.java#L1019): Make this anonymous inner class a lambda [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1604)
1. ![MAJOR][MAJOR] [hot.js#L51](https://github.com/TGAC/miso-lims/blob/73862dfebeda7ae6c1c1fc7017ed296a57eaa7b8/miso-web/src/main/webapp/scripts/hot.js#L51): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)
1. ![MAJOR][MAJOR] [hot.js#L238](https://github.com/TGAC/miso-lims/blob/73862dfebeda7ae6c1c1fc7017ed296a57eaa7b8/miso-web/src/main/webapp/scripts/hot.js#L238): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)
1. ![MAJOR][MAJOR] [hot.js#L379](https://github.com/TGAC/miso-lims/blob/73862dfebeda7ae6c1c1fc7017ed296a57eaa7b8/miso-web/src/main/webapp/scripts/hot.js#L379): Define this function outside of a loop. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AFunctionDefinitionInsideLoop)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**